### DISCUSSION TITLE
e2e protection configurations for all the profiles

### DISCUSSION THREAD

**Discussion by NagarjunaReddyG5262** [COVESA/vsomeip#771] [Q&A] [answered]
Hi 

I am trying to use e2e protection in my implementation and I need to use all the available autosar e2e protection profiles, but here https://github.com/COVESA/vsomeip/tree/master/test/network_tests/e2e_tests/conf  I can only find examples of vsomeip json configuration files for Profile 04 and Profile 07, could someone help me to understand the json configuration for all the available autosar profiles like 1,2,4,5,6,7,11,22. Thanks in advance.

**Comment by fcmonteiro** [comment, marked as answer]
Hi @NagarjunaReddyG5262, here's the part of our E2E configuration.
Thanks

## 20. E2E

- **e2e** - Used to configure the E2E protection for the specified events
    - **e2e_enabled** - Specifies if E2E protection should be enabled or disabled. Use `true` to enable.
    - **protected** - Specify the protected events
        - **service_id** - Specifies the service ID.
        - **event_id** - Specifies the event ID to be protected.
        - **variant** -  Specifies if it should protect (send) or check (receive) the messages, or both. Possible values are `protector`, `checker` and `both`, respectively.
        - **profile** - Specify the E2E profile to be used. Valid values are `CRC8` for Profile 01, `P04` for Profile 04, `P07` for Profile 07 and `CRC32` for the Custom Profile with ethernet CRC.
            - Profile 01
                - **data_id** - Specifies a system wide unique 16 bit numerical identifier.
                - **crc_offset** - Specifies the offset of the CRC in bytes.
                - **data_length** - Specifies the length of all data in bits.
                - **counter_offset** - Specifies the offset of the counter in bits. Default value is `8`.
                - **data_id_nibble_offset** - Specifies the offset of the dataID nibble. Default value is `12`.
                - **data_id_mode** - Specifies the dataID mode (valid values are `0`, `1`, `2` and `3`). It impacts which part of the dataID is used in CRC calculation and if any port of the dataID is sent in the E2E Header.
            - Profile 04
                - **data_id** - Specifies a system wide unique 32 bit numerical identifier.
                - **crc_offset** - Specifies the offset of the E2E header in bits. It must start at 64 (8 bits) and be a multiple of 8.
                - **min_data_length** - Specifies the minimum length of the data in bits. Default value is `0`.
                - **max_data_length** - Specifies the maximum length of the data in bits. Default value is `0xFFFF`.
                - **max_delta_counter** - Specifies the maximum allowed difference between the counter value of the current message and the previous valid message. Default value is `0xFFFF`.
            - Profile 07
                - **data_id** - Specifies a system wide unique 32 bit numerical identifier.
                - **crc_offset** - Specifies the offset of the E2E header in bits. It must start at 64 (8 bits) and be a multiple of 8.
                - **min_data_length** - Specifies the minimum length of the data in bits. Default value is `0`.
                - **max_data_length** - Specifies the maximum length of the data in bits. Default value is `0xFFFFFFFF`.
                - **max_delta_counter** - Specifies the maximum allowed difference between the counter value of the current message and the previous valid message. Default value is `0xFFFFFFFF`.
            - Custom
                - **crc_offset** - Specifies the offset of the CRC (same as E2E Header in this profile) in bytes.

<details><summary>Simple E2E configuration</summary>

```
"e2e" :
    {
        "e2e_enabled" : "true",
        "protected" :
        [
            {
                "service_id" : "0x1507",
                "event_id" : "0x01",
                "profile" : "CRC32",
                "variant" : "protector",
                "crc_offset" : "0"
            },
            {
                "service_id" : "0x1507",
                "event_id" : "0x8001",
                "profile" : "CRC8",
                "variant" : "checker",
                "crc_offset" : "0",
                "data_id_mode" : "3",
                "data_length" : "56",
                "data_id" : "0xA73"
            },
            {
                "service_id" : "0x1507",
                "event_id" : "0x02",
                "profile" : "CRC8",
                "variant" : "checker",
                "crc_offset" : "0",
                "data_id_mode" : "3",
                "data_length" : "56",
                "data_id" : "0xA73"
            }
        ]
    },
```
</details>

<details><summary>E2E Profile 01 configuration</summary>

**Server-side**

```
"e2e" :
{
    "e2e_enabled" : "true",
    "protected" :
    [
        {
            "service_id" : "0x1234",
            "event_id" : "0x80e9",
            "profile" : "CRC8",
            "variant" : "protector",
            "crc_offset" : "0",
            "data_id_mode" : "3",
            "data_length" : "24",
            "data_id" : "0x1FF"
        },
        {
            "service_id" : "0x1234",
            "event_id" : "0x80ea",
            "profile" : "CRC8",
            "variant" : "protector",
            "crc_offset" : "0",
            "data_id_mode" : "3",
            "data_length" : "24",
            "data_id" : "0x1FF"
        }
    ]
},
```

**Client-side**

```
"e2e" :
{
    "e2e_enabled" : "true",
    "protected" :
    [
        {
            "service_id" : "0x1234",
            "event_id" : "0x80e9",
            "profile" : "CRC8",
            "variant" : "checker",
            "crc_offset" : "0",
            "data_id_mode" : "3",
            "data_length" : "24",
            "data_id" : "0x1FF"
        },
        {
            "service_id" : "0x1234",
            "event_id" : "0x80ea",
            "profile" : "CRC8",
            "variant" : "checker",
            "crc_offset" : "0",
            "data_id_mode" : "3",
            "data_length" : "24",
            "data_id" : "0x000"
        }
    ]
},
```

</details>

<details><summary>E2E Profile 04 configuration</summary>

**Server-side**

```
"e2e" :
{
    "e2e_enabled" : "true",
    "protected" :
    [
        {
            "service_id" : "0x1234",
            "event_id" : "0x80e9",
            "profile" : "P04",
            "variant" : "protector",
            "crc_offset" : "64",
            "data_id" : "0xFF"
        },
        {
            "service_id" : "0x1234",
            "event_id" : "0x80ea",
            "profile" : "P04",
            "variant" : "protector",
            "crc_offset" : "64",
            "data_id" : "0xFF"
        }
    ]
},
```

**Client-side**

```
"e2e" :
{
    "e2e_enabled" : "true",
    "protected" :
    [
        {
            "service_id" : "0x1234",
            "event_id" : "0x80e9",
            "profile" : "P04",
            "variant" : "checker",
            "crc_offset" : "64",
            "data_id" : "0xFF"
        },
        {
            "service_id" : "0x1234",
            "event_id" : "0x80ea",
            "profile" : "P04",
            "variant" : "checker",
            "crc_offset" : "64",
            "data_id" : "0x00"
        }
    ]
},
```

</details>
